### ISSUE TITLE
[question] website documentaton

### ISSUE THREAD

**Issue by jasonLaster** [microsoft/monaco-editor#190]
I'd like to evaluate using monaco in the firefox debugger, but it is difficult to get started without project documentation.

Here are a couple questions to get the conversation started:
- _where is the monaco-editor-code?_ - it took me some digging to find the monaco-editor-core package.json and track down the src/vs/editor directory
- _what is the public api?_ - it would be nice to use a tool like documentation.js to document the public api. Codemirror has a well defined api and usage page.
- _how do i do x?_  - codemirror has great small demos which show how i can extend the editor, which could be useful here.

I don't mention this to be a thorn, but i'm genuinely curious. I'd be happy to help setup docs as well and be a constructive citizen too :)

**Comment by alexandrudima** [comment]
Hi Jason,

I'm sorry things are not obvious, and a bit complicated, mostly due to our history. The Monaco Editor grew up inside the Monaco Workbench, which later became VSCode. 

So the source code of the editor still lives inside vscode at https://github.com/Microsoft/vscode/tree/master/src/vs/editor . There is a lot of source code shared between the editor and VSCode, especially in `base`.

From the vscode repository we publish to npm as [`monaco-editor-core`](https://www.npmjs.com/package/monaco-editor-core) two bundled versions of the editor (one minified and one non-minified), accompanied by a monaco.d.ts that describes our supported API. We use AMD and pull in only the subset of modules needed by the standalone editor, modules from:
* https://github.com/Microsoft/vscode/tree/master/src/vs/base
* https://github.com/Microsoft/vscode/tree/master/src/vs/platform
* https://github.com/Microsoft/vscode/tree/master/src/vs/editor
* https://github.com/Microsoft/vscode/tree/master/src/vs/languages

The main AMD entry point is https://github.com/Microsoft/vscode/blob/master/src/vs/editor/editor.main.ts

Our vision is that  `monaco-editor-core` is supposed to be just the editor and its UI (language agnostic), but at this point also ships support for the following languages: html, php, handlebars and razor. The reason is because VSCode also uses the exact same languages. We are working towards migrating these old implementations to VSCode extensions, and we are quite close to do so, perhaps this month or the next one we will be done. We will then be free to make `monaco-editor-core` contain just the editor (language agnostic).

We have already cleaned up the vast majority of the language supports (in the form of monaco editor core "plugins") and they live:
* https://github.com/Microsoft/monaco-languages => npm `monaco-languages` -- here is a collection of many languages that mostly contribute colorization and basic language configuration (i.e. what is the comment token such that `ctrl+/` can work, etc).
* https://github.com/Microsoft/monaco-css => npm `monaco-css` -- here live rich language support for CSS, SCSS and LESS -- think of it like validation, intellisense, find all references, etc. They are built by using `monaco-editor-core` API that helps them start and manage web workers with synchronized buffers, etc.
* https://github.com/Microsoft/monaco-json => npm `monaco-json` -- rich language support for JSON
* https://github.com/Microsoft/monaco-typescript => npm `monaco-typescript` -- rich language support for JavaScript and TypeScript based on the TypeScript language service

In this repository we have the scripts that pull things together and create a distribution that contains the `monaco-editor-core` + all these languages in the form of the `monaco-editor`. 

The `monaco-editor`, once you download it from npm, does not consist of a single script due to our desire to load code only when it is needed. e.g. If you create a buffer (model) that is `css`, we want to load only the necessary source code for css support (`vs/language/css/cssMode` that then decides to spin up a web worker with `vs/language/css/cssWorker`) and not load all the rest of the languages, thus we rely on the integrator to provide an AMD loader, or they can use the one we ship (`loader.js`).

We also have multiple scripts because we don't want to load Korean strings unless you configure the editor to start up in Korean and due to how web workers must load code independently.

Documentation is lacking, and I am sorry about that, my main day-to-day responsibilities lie within VSCode, but I try my best to document things in the form of samples:
* samples playground - https://microsoft.github.io/monaco-editor/playground.html
* colorization playground - https://microsoft.github.io/monaco-editor/monarch.html

There are more advanced use-cases that are lacking in the samples, especially around how to work with multiple buffers. Long story short:
* create one editor (`monaco.editor.create`)
* create multiple models (`monaco.editor.createModel`)
* a model has a language id associated with it
* an editor can have one model at a time, a model can be in two different editors at the same time
* use `setModel(model)`, `setModel(null)`, `saveViewState()`, `restoreViewState()` to implements tabs
* don't forget to `dispose` models and editors when you don't need them anymore

I plan to write some tutorial(s) when I get a chance, in the meantime we can use issues + search for questions.

**TL;DR**;
* source code for editor core: https://github.com/Microsoft/vscode/tree/master/src/vs
* source code for specific languages:
  * https://github.com/Microsoft/monaco-languages
  * https://github.com/Microsoft/monaco-css
  * https://github.com/Microsoft/monaco-json
  * https://github.com/Microsoft/monaco-typescript
* public API is described in `monaco.d.ts` that comes inside the `monaco-editor` npm module (such that it can be versioned correctly) -- also latest version available at https://github.com/Microsoft/monaco-editor/blob/master/website/playground/monaco.d.ts.txt
* we have a bunch of how to do in the playground section at https://microsoft.github.io/monaco-editor/playground.html -- just use the dropdown.

**Comment by jasonLaster** [comment]
Thanks @alexandrudima this is a lot of good information on how to move forward.
